tpu-client-next/src/connection_workers_scheduler.rs

```diff
@@ -0,0 +1,121 @@
+//! This module defines [`ConnectionWorkersScheduler`] which creates and
+//! orchestrates `ConnectionWorker` instances.
+use {
+    super::{leader_updater::LeaderUpdater, SendTransactionStatsPerAddr},
+    crate::{
+        connection_worker::ConnectionWorker,
+        quic_networking::{
+            create_client_config, create_client_endpoint, QuicClientCertificate, QuicError,
+        },
+        transaction_batch::TransactionBatch,
+        workers_cache::{WorkerInfo, WorkersCache, WorkersCacheError},
+    },
+    log::*,
+    quinn::Endpoint,
+    solana_sdk::signature::Keypair,
+    std::{net::SocketAddr, sync::Arc},
+    thiserror::Error,
+    tokio::sync::mpsc,
+};
+
+/// Size of the channel to transmit transaction batches to the target workers.
+const WORKER_CHANNEL_SIZE: usize = 2;
+
+/// [`ConnectionWorkersScheduler`] is responsible for managing and scheduling
+/// connection workers that handle transactions over the network.
+pub struct ConnectionWorkersScheduler;
+
+#[derive(Debug, Error, PartialEq)]
+pub enum ConnectionWorkersSchedulerError {
+    #[error(transparent)]
+    QuicError(#[from] QuicError),
+    #[error(transparent)]
+    WorkersCacheError(#[from] WorkersCacheError),
+    #[error("Leader receiver unexpectedly dropped.")]
+    LeaderReceiverDropped,
+}
+
+impl ConnectionWorkersScheduler {
+    /// Runs the main loop that handles worker scheduling and management for
+    /// connections. Returns the error quic statistics per connection address or
+    /// an error if something goes wrong. Importantly, if some transactions were
+    /// not delivered due to network problems, they will not be retried when the
+    /// problem is resolved.
+    pub async fn run(
+        bind: SocketAddr,
+        validator_identity: Option<Keypair>,
+        mut leader_updater: Box<dyn LeaderUpdater>,
+        mut transaction_receiver: mpsc::Receiver<TransactionBatch>,
+        num_connections: usize,
+    ) -> Result<SendTransactionStatsPerAddr, ConnectionWorkersSchedulerError> {
+        let endpoint = Self::setup_endpoint(bind, validator_identity)?;
+        debug!("Client endpoint bind address: {:?}", endpoint.local_addr());
+        let mut workers = WorkersCache::new(num_connections);
+        loop {
+            let Some(transaction_batch) = transaction_receiver.recv().await else {
```

Comment:
Would there be performance benefit to pulling all transaction batches off the receiver?

Reply:
Do you mean take all the batches from `transaction_receiver` and send them all after that? 
The idea is that the bounded channel is used here (of size 2 in case of transaction-bench), so that we have a back-pressure on the sender side of the channel. If we take all the batches from the receiver, the generator part (which puts batches to sender) will create some new transactions although we maybe are slow in sending.